### ISSUE TITLE
React Native Support

### ISSUE THREAD

**Issue by blackxored** [layerhq/layer-websdk#85]
I would love if this library officially supported React Native.

Here are my findings so far:

The easy parts:
- Need to add `fs`, `vm`, `stream` and `filereader` as deps in this package's `package.json`. If running on Node this shouldn't affect normal `require` calls. Needs to test browserify et al.
- #84 needs to be merged, otherwise everything would load, but Layer's async funcs won't work when run on Chrome debugger or maybe even device. 

The hard parts:
- On message-part.js the check for FileReader seems to be lackluster (investigating this), failing on JavaScript Core.
- External node modules deps need `crypto`. Need to devise an option to override to use `react-native-crypto` instead. So far:  `get-random-bytes/index.js`, `sjcl/sjcl.js` on dev. 
  `brorand/index.js` while bundling (might be specific to my project?). Unfortunately I'm not aware of an easy fix for this one, there might be a react-native keyword in package.json in the future, but not sure it would help.
- Polyfill `online-state-manager`'s connectivity checks.

Will update this if I gather new info. Looking forward to your comments.

**Comment by michaelkantor** [comment]
My take on this:
1. See our [layer-react library](https://www.npmjs.com/package/layer-react); this a module whose original designer had intended it to run in React Native (we removed support for this when he left as we don't have sufficient familiarity/confidence to support it right now)
2. There has been some effort to get this to run in node in the past... which was put aside for the sake of getting this to ship.  Evidence of this remains: https://github.com/layerhq/layer-websdk/blob/master/package.json#L27-L33
3. Crypto is NOT a dependency in the current beta release: https://github.com/layerhq/layer-websdk/releases/tag/2.0.0-beta.1
4. Polyfill for connectivity checks is worth investigating

While I can't say if all changes you make are welcome or not until I review them, I do look forwards to reviewing them. However, you should focus your efforts on the Beta; changes made to master probably won't survive the transition to `v2.0`.  Branch name is `dev1.1`.

**Comment by blackxored** [comment]
Hi, thanks for the answer!

1. I am indeed using layer-react but it requires a `Client` from this package. 
2. RN should support `browser` keyword in `package.json` or even `react-native`, but the problem seems to be the `require` calls across the codebase; and the checks for `typeof(window)` aren't enough to differentiate the env.
3. I can't quite afford to go the beta route just now :( got an app to launch 🎊 
4. `NetworkInfo` should do it, but not sure on how to reliably detect RN (`navigator.product` is still Gecko) or whether it could be able to be used at the time layer registers its event listener for it.

**Comment by michaelkantor** [comment]
Ok, so...

1. Detecting RN: Perhaps if no other means exists, we require a property be set on `window` as part of initialization of any RN app?
2. Ok, app to launch: focus on getting a fork to work, and we'll sync up on making that work in the WebSDK 2.0 later.  For your fork, you can get rid of Crypto and substitute any code that generates something that looks like a UUID; does not need to be a valid UUID, just correctly formatted.  That UUID is NOT sent to the server (not until WebDK 2.0).  This one line in the beta generates a UUID: https://github.com/layerhq/layer-websdk/blob/dev1.1/src/client-utils.js#L17

**Comment by jcham** [comment]
Any update on this, I would also love to see a React Native compatible implementation. 
If that's not coming any time soon, any more details on how to get it to run, even if with reduced functionality?

**Comment by michaelkantor** [comment]
We have not yet had time to do serious experimentation with React Native; I'm hoping that project can get prioritized for sometime next year.  For earlier feedback, there are a few github issues where people have investigated... try @jcham mentioning them to see if anyone has worked out any solutions or forks they are willing to share.

**Comment by Monte9** [comment]
@michaelkantor any official updates from Layer on this project?

@blackxored did you have any success implementing Layer for React Native on your own? I am looking to at least build a simple dev version, doesn't have to be production ready yet. Any working forks that you could link me to? Just need to show proof of concept to show Layer implemented with React Native. Thanks in advance.

**Comment by michaelkantor** [comment]
No official updates @Monte9 however, unofficially, we recently hired a developer with React Native background, and are hoping to have a React Native sample app and better WebSDK support in the next few months.

**Comment by digitaldavenyc** [comment]
As a suggestion, I would make this a high priority. Competitors to Layer have already added support for React-Native and companies with existing React-Native code-bases will have to go with another vendor for messaging.

**Comment by michaelkantor** [comment]
Layer WebSDK 3.2.3 contains improvements around working with React Native.  You must change your import/require code a little to use this:

```javascript
var layer = require('layer-websdk/index-react-native');
```
or
```javascript
import Layer from 'layer-websdk/index-react-native';
```

Note that Blobs are not currently handled, which will prevent sending of Rich Content.  It looks like this is something being worked on by the React Native team itself, so we are holding off on crazy Blob hacks for now.

React Native Apps have been tested by @SpecialDOS.

**Comment by adylevy** [comment]
Hi @michaelkantor 
Looks like this commit and line broke support for RN - 
https://github.com/layerhq/layer-websdk/commit/0196f8df99546cd9b179ba75c7450cd1a5a40da4#diff-e1b696fce5897dfa7d19f7390fa96e91R126

atob is not supported in RN

**Comment by michaelkantor** [comment]
Verified @adylevy; unfortunately, RN support is not maintained as a first class citizen (i.e. no unit tests for it).

Please file this as a separate ticket/github-issue. I'll see that this gets out with our next update (could be a few weeks before that ships).

**Comment by michaelkantor** [comment]
Fix is in 3.4.0 @adylevy
